src/Couchbase.Lite.Shared/API/Database/Database.cs

```diff
@@ -1265,40 +1250,62 @@ private void Save([NotNull]Document doc, C4Document** outDoc, C4Document* baseDo
+        private void SaveResolvedDocument([CanBeNull]Document resolvedDoc, [NotNull]Document localDoc, [NotNull]Document remoteDoc)
         {
-            if (!ReferenceEquals(resolved, localDoc)) {
-                resolved.Database = this;
+            if (resolvedDoc == null) {
+                if (localDoc.IsDeleted)
+                    resolvedDoc = localDoc;
+
+                if (remoteDoc.IsDeleted)
+                    resolvedDoc = remoteDoc;
             }
 
-            // The remote branch has to win, so that the doc revision history matches the server's
+            if (resolvedDoc!=null && !ReferenceEquals(resolvedDoc, localDoc)) {
+                resolvedDoc.Database = this;
+            }
+
+            // The remote branch has to win, so that the doc revision history matches the server's.
             var winningRevID = remoteDoc.RevID;
             var losingRevID = localDoc.RevID;
 
             FLSliceResult mergedBody = (FLSliceResult)FLSlice.Null;
-            if (!ReferenceEquals(resolved, remoteDoc)) {
-                // Unless the remote revision is being used as-is, we need a new revision:
-                try {
-                    mergedBody = resolved.Encode();
-                } catch (ObjectDisposedException) {
-                    WriteLog.To.Sync.E(Tag, "Resolved document for {0} somehow got disposed!",
-                        new SecureLogString(resolved.Id, LogMessageSensitivity.PotentiallyInsecure));
-                    throw new RuntimeException(
-                        "Resolved document was disposed before conflict resolution completed.  Please file a bug report at https://github.com/couchbase/couchbase-lite-net");
+            C4RevisionFlags mergedFlags = 0;
+            if (!ReferenceEquals(resolvedDoc, remoteDoc)) {
+                var isDeleted = true;
+                if (resolvedDoc != null) {
+                    // Unless the remote revision is being used as-is, we need a new revision:
+                    try {
+                        mergedBody = resolvedDoc.Encode();
+                    } catch (ObjectDisposedException) {
+                        WriteLog.To.Sync.E(Tag, "Resolved document for {0} somehow got disposed!",
+                            new SecureLogString(resolvedDoc.Id, LogMessageSensitivity.PotentiallyInsecure));
+                        throw new RuntimeException(
+                            "Resolved document was disposed before conflict resolution completed.  Please file a bug report at https://github.com/couchbase/couchbase-lite-net");
+                    }
+                    if (mergedBody.Equals((FLSliceResult)FLSlice.Null))
+                        throw new RuntimeException("Resolved document contains a null body");
+                    isDeleted = resolvedDoc.IsDeleted;
+                } else if (resolvedDoc == null || resolvedDoc.IsEmpty) {
```

Comment:
Looks like this check doesn't make as much sense as I thought.  This is an `else` against `resolvedDoc != null` so it is guaranteed to be `null` here.  I guess just a simple `else` will do.